Sources/NIO/ChannelPipeline.swift

```diff
@@ -162,19 +162,136 @@ public final class ChannelPipeline: ChannelInvoker {
+            self.add0(name: name, handler: handler, relativeHandler: after, operation: self.add0(context:after:), promise: promise)
+        } else {
+            self.eventLoop.execute {
+                self.add0(name: name, handler: handler, relativeHandler: after, operation: self.add0(context:after:), promise: promise)
             }
         }
+
         return promise.futureResult
     }
 
-    /// Add the handler to the `ChannelPipeline`. This operation must only be called when within the `EventLoop` thread.
-    private func add0(name: String?, handler: ChannelHandler, first: Bool, promise: EventLoopPromise<Void>) {
+    /// Add a `ChannelHandler` to the `ChannelPipeline` immediately before a `ChannelHandler` that is already present
+    /// in the `ChannelPipeline`.
+    ///
+    /// - parameters:
+    ///     - name: The name to use for the `ChannelHandler` when its added. If none is specified, a name will be
+    ///         automatically generated.
+    ///     - handler: The `ChannelHandler` to add.
+    ///     - after: The pre-existing `ChannelHandler` that `handler` should be inserted immediately before.
+    /// - returns: An `EventLoopFuture` that will be notified when the `ChannelHandler` is added.
+    public func add(name: String? = nil, handler: ChannelHandler, before: ChannelHandler) -> EventLoopFuture<Void> {
+        let promise: EventLoopPromise<Void> = self.eventLoop.newPromise()
+
+        if self.eventLoop.inEventLoop {
+            self.add0(name: name, handler: handler, relativeHandler: before, operation: self.add0(context:before:), promise: promise)
+        } else {
+            self.eventLoop.execute {
+                self.add0(name: name, handler: handler, relativeHandler: before, operation: self.add0(context:before:), promise: promise)
+            }
+        }
+
+        return promise.futureResult
+    }
+
+    /// Synchronously add a `ChannelHandler` to the pipeline, relative to another `ChannelHandler`,
+    /// where the insertion is done by a specific operation.
+    ///
+    /// May only be called from on the event loop.
+    ///
+    /// This will search the pipeline for `relativeHandler` and, if it cannot find it, will fail
+    /// `promise` with `ChannelPipelineError.notFound`.
+    ///
+    /// - parameters:
+    ///     - name: The name to use for the `ChannelHandler` when its added. If none is specified, a name will be
+    ///         automatically generated.
+    ///     - handler: The `ChannelHandler` to add.
+    ///     - relativeHandler: The `ChannelHandler` already in the `ChannelPipeline` that `handler` will be
+    ///         inserted relative to.
+    ///     - operation: A callback that will insert `handler` relative to `relativeHandler`.
+    ///     - promise: An `EventLoopPromise<Void>` that will fire when the operation is complete, or will fire with
+    ///         an error if it could not be completed.
+    private func add0(name: String?,
```

Comment:
I'm curious, what is the `0` suffix intended to denote? To me they read like JNI natives!?

Reply:
The convention, weakly enforced though it is, is that the `0` suffix indicates the actual implementation of a method that needs to be dispatched to an event loop thread. That is, we have a `foo0` method whenever there is a `foo` that must be thread-safe.

Reply:
I usually use `0` for private methods that also exists with the same name as public.

Reply:
Thanks.